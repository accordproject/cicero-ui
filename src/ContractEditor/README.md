## Cicero-UI - ContractEditor

### Usage

```
npm install @accordproject/cicero-ui
```

```
import { ContractEditor } from '@accordproject/cicero-ui';

const clausePropsObject = {
    BODY_FONT (string),
    CLAUSE_BACKGROUND (string),
    CLAUSE_BORDER (string),
    CLAUSE_DELETE (string),
    CLAUSE_DELETE_FUNCTION (function),
    HEADER_FONT (string),
}

const editorPropsObject = {
    BUTTON_BACKGROUND_INACTIVE (string),
    BUTTON_BACKGROUND_ACTIVE (string),
    BUTTON_SYMBOL_INACTIVE (string),
    BUTTON_SYMBOL_ACTIVE (string),
    DROPDOWN_COLOR (string),
    TOOLBAR_BACKGROUND (string),
    TOOLTIP_BACKGROUND (string),
    TOOLTIP (string),
    TOOLBAR_SHADOW (string),
    WIDTH (string),
}

function storeLocal(editor) {
    localStorage.setItem('markdown-editor', editor.getMarkdown());
}

ReactDOM.render(<ContractEditor 
    clauseProps={clauseProps(props.removeFromContract)}
    editorProps={editorPropsObject}
    onChange={storeLocal} />,
  document.getElementById('root'));
```

### Props

- `value` : An `object` which is the initial contents of the editor.
- `onChange` : A callback `function` called when the contents of the editor change.
- `lockText` : A `boolean` to lock all non variable text.
- `loadTemplateObject` : A callback `function` to load a template.
- `parseClause` : A callback `function` to parse the contents of a clause.
- `templates` : An `array` which contains template objects with the following keys: `uir`, `name`, `version`, `description`.
- `editorProps` : An optional `object` for the contract editor styling passed down through an object to Markdown Editor, see below.
- `clauseProps` : An `object` for the clauses in the editor which contains a deletion function, see below.


### Specifications

`editorProps`:
You can style the toolbar of this components, as well as the width of the editor. An object may be passed down this component which will then be picked up by our `markdown-editor`, with the following possible CSS inputs as strings:
- `BUTTON_BACKGROUND_INACTIVE`
- `BUTTON_BACKGROUND_ACTIVE`
- `BUTTON_SYMBOL_INACTIVE`
- `BUTTON_SYMBOL_ACTIVE`
- `DROPDOWN_COLOR`
- `TOOLBAR_BACKGROUND`
- `TOOLTIP_BACKGROUND`
- `TOOLTIP`
- `TOOLBAR_SHADOW`
- `WIDTH`

`clauseProps`:
You can style the Clause Components within the `ContractEditor`. An object may be passed down this component with the following possible CSS inputs as strings, as well as a deletion function:
- `BODY_FONT`
- `CLAUSE_BACKGROUND`
- `CLAUSE_BORDER`
- `CLAUSE_DELETE`
- `CLAUSE_DELETE_FUNCTION`
- `HEADER_FONT`