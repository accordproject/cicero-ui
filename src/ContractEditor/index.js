/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { MarkdownEditor } from '@accordproject/markdown-editor';
import List from '@accordproject/markdown-editor/dist/plugins/list';
import ClausePlugin from './plugins/clausePlugin';

/**
 * A rich text contract editor
 */
class ContractEditor extends React.Component {
  constructor(props) {
    super(props);
    this.props = props;
    this.props.plugins.push(List(), ClausePlugin());
  }

  /**
   * Render this React component
   * @return {*} the react component
   */
  render() {
    return (<MarkdownEditor markdown={this.props.markdown}
      onChange={this.props.onChange}
      plugins={this.props.plugins}/>);
  }
}

/**
 * The property types for this component
 */
ContractEditor.propTypes = {
  markdown: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  plugins: PropTypes.arrayOf(PropTypes.shape({
    onEnter: PropTypes.func,
    onKeyDown: PropTypes.func,
    renderNode: PropTypes.func.isRequired,
    toMarkdown: PropTypes.func.isRequired,
    fromMarkdown: PropTypes.func.isRequired,
    fromHTML: PropTypes.func.isRequired,
    plugin: PropTypes.string.isRequired,
    tags: PropTypes.arrayOf(PropTypes.string).isRequired,
    markdownTags: PropTypes.arrayOf(PropTypes.string).isRequired,
    schema: PropTypes.object.isRequired,
  })),
};

export { ContractEditor };
