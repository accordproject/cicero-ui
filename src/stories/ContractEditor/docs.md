# Contract Editor

#### Notes for Contract Editor

---

Supports code snippets too:

```jsx
<div>
  Foo
</div>
```