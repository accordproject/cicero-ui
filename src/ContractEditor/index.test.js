import React from 'react';
import { render } from '@testing-library/react';
import { SlateTransformer } from '@accordproject/markdown-slate';
import { Chance } from 'chance';
import ContractEditor from './index';

const slateTransformer = new SlateTransformer();
const chance = new Chance();

const props = {
  value: null,
  onChange: () => 1,
  lockText: true,
  template: {},
  loadTemplateObject: () => 1,
  parseClause: () => 1,
  editorProps: {
    BUTTON_BACKGROUND_INACTIVE: null,
    BUTTON_BACKGROUND_ACTIVE: null,
    BUTTON_SYMBOL_INACTIVE: null,
    BUTTON_SYMBOL_ACTIVE: null,
    DROPDOWN_COLOR: null,
    TOOLBAR_BACKGROUND: null,
    TOOLTIP_BACKGROUND: null,
    TOOLTIP: null,
    TOOLBAR_SHADOW: null,
    WIDTH: '600px',
  }
};

const setup = (content = null) => {
  const ref = React.createRef();
  const value = content ? slateTransformer.fromMarkdown(content) : null;
  return render(<ContractEditor {...props} value={value} ref={ref} />);
};

beforeAll(() => {
  window.getSelection = () => ({
    removeAllRanges: () => {}
  });

  global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
      nodeName: 'BODY',
      ownerDocument: document,
    },
  });
});

describe('<ContractEditor />', () => {
  describe('on initialization', () => {
    it('renders page correctly', () => {
      const { baseElement } = setup();
      expect(baseElement).toMatchSnapshot();
    });

    // it('should render specified value', () => {
    //   const content = chance.sentence();
    //   const container = setup(content);
    //   const textbox = container.getByRole('textbox');
    //   expect(textbox.innerHTML).toEqual(expect.stringContaining(content));
    // });
  });

  describe('on formatting', () => {
    it('should render hyperlink properly', async () => {
      const text = chance.sentence({ words: 2 });
      const url = chance.url();
      setup(`[${text}](${url})`);
      const link = document.querySelector(`a[href='${url}']`);
      expect(link).toBeTruthy();
      expect(link.textContent).toBe(text);
      expect(link.href).toBe(url);
    });

    it('should render bold properly', () => {
      const text = chance.sentence({ words: 2 });
      setup(`**${text}**`);
      expect(document.querySelector('strong').textContent).toBe(text);
    });

    it('should render italic properly', () => {
      const text = chance.sentence({ words: 2 });
      setup(`*${text}*`);
      expect(document.querySelector('em').textContent).toBe(text);
    });

    it('should render code properly', () => {
      const text = chance.hashtag();
      setup(`\`${text}\``);
      expect(document.querySelector('code').textContent).toBe(text);
    });

    it('should render blockquote properly', () => {
      const fisrtSentence = chance.sentence();
      const secondSentence = chance.sentence();
      const container = setup(`> ${fisrtSentence}\n> ${secondSentence}`);
      expect(document.querySelector('blockquote').textContent).toBe(fisrtSentence + secondSentence);
    });
  });
});
