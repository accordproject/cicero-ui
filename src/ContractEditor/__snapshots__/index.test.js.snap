// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContractEditor /> on initialization renders page correctly 1`] = `
<body>
  <div>
    <div
      class="ap-markdown-editor"
    >
      <div
        class="sc-EHOje gvyJFh"
        id="slate-toolbar-wrapper-id"
      >
        <div
          class="sc-bdVaJa jgHvjx format-toolbar"
        >
          <div
            aria-expanded="false"
            class="ui simple dropdown toolbar-0x0"
            role="listbox"
            style="color: rgb(65, 79, 88); align-self: center;"
            tabindex="0"
          >
            <div
              aria-atomic="true"
              aria-live="polite"
              class="text"
              role="alert"
            >
              Style
            </div>
            <i
              aria-hidden="true"
              class="dropdown icon"
            />
            <div
              class="menu transition"
            >
              <div
                class="item"
                role="option"
              >
                <span
                  class="text"
                >
                  Normal
                </span>
              </div>
              <div
                class="item"
                role="option"
                style="font-size: 25px; line-height: 23px; font-weight: bold; color: rgb(18, 35, 48);"
              >
                <span
                  class="text"
                >
                  Header 1
                </span>
              </div>
              <div
                class="item"
                role="option"
                style="font-size: 20px; line-height: 20px; font-weight: bold; color: rgb(18, 35, 48);"
              >
                <span
                  class="text"
                >
                  Header 2
                </span>
              </div>
              <div
                class="item"
                role="option"
                style="font-size: 16px; line-height: 16px; font-weight: bold; color: rgb(18, 35, 48);"
              >
                <span
                  class="text"
                >
                  Header 3
                </span>
              </div>
            </div>
          </div>
          <div
            class="sc-htpNat jqGvzC toolbar-4x0"
          />
          <svg
            aria-label="Bold (Ctrl+B)"
            class="sc-bwzfXH jKgcJe toolbar-0x1"
            height="25px"
            viewBox="0 0 11 13"
            width="25px"
          >
            <g
              fill="none"
              fill-rule="evenodd"
              font-family="IBMPlexSans-Bold, IBM Plex Sans"
              font-size="18"
              font-weight="bold"
              stroke="none"
              stroke-width="1"
            >
              <g
                fill="#949CA2"
                transform="translate(-440.000000, -48.000000)"
              >
                <text>
                  <tspan
                    x="439"
                    y="61"
                  >
                    B
                  </tspan>
                </text>
              </g>
            </g>
          </svg>
          <svg
            aria-label="Italic (Ctrl+I)"
            class="sc-bwzfXH jKgcJe toolbar-0x2"
            height="25px"
            viewBox="0 0 10 13"
            width="25px"
          >
            <g
              fill="none"
              fill-rule="evenodd"
              font-family="IBMPlexSans-BoldItalic, IBM Plex Sans"
              font-size="18"
              font-style="italic"
              font-weight="bold"
              stroke="none"
              stroke-width="1"
            >
              <g
                fill="#949CA2"
                transform="translate(-477.000000, -48.000000)"
              >
                <text>
                  <tspan
                    x="478"
                    y="61"
                  >
                    I
                  </tspan>
                </text>
              </g>
            </g>
          </svg>
          <div
            class="sc-htpNat jqGvzC toolbar-4x1"
          />
          <svg
            aria-label="Code (Ctrl+Alt+C)"
            class="sc-bwzfXH fDMMRC toolbar-1x0"
            height="25px"
            viewBox="0 0 20 12"
            width="25px"
          >
            <g
              fill="none"
              fill-rule="evenodd"
              stroke="none"
              stroke-width="1"
            >
              <g
                fill="#949CA2"
                fill-rule="nonzero"
                transform="translate(-508.000000, -49.000000)"
              >
                <g
                  transform="translate(508.000000, 49.000000)"
                >
                  <path
                    d="M19.609,6.787 C19.843431,6.5527408 19.9751474,6.23491402 19.9751474,5.9035 C19.9751474,5.57208598 19.843431,5.2542592 19.609,5.02 L15.609,1.02 C15.2952486,0.694966717 14.8305318,0.564516169 14.3934634,0.678786996 C13.956395,0.793057823 13.6149929,1.13426688 13.5004748,1.57127054 C13.3859567,2.0082742 13.5161443,2.47306473 13.841,2.787 L16.6,5.55 C16.795191,5.74524993 16.795191,6.06175007 16.6,6.257 L13.841,9.02 C13.3674216,9.51060731 13.3744161,10.2902393 13.8567205,10.7722709 C14.3390249,11.2543025 15.1186607,11.2608559 15.609,10.787 L19.609,6.787 Z"
                  />
                  <path
                    d="M6.609,1.02 C6.37470042,0.785262804 6.05665923,0.653351971 5.725,0.653351971 C5.39334077,0.653351971 5.07529958,0.785262804 4.841,1.02 L0.841,5.02 C0.606569022,5.2542592 0.474852621,5.57208598 0.474852621,5.9035 C0.474852621,6.23491402 0.606569022,6.5527408 0.841,6.787 L4.841,10.787 C5.15475142,11.1120333 5.61946823,11.2424838 6.05653661,11.128213 C6.49360499,11.0139422 6.83500715,10.6727331 6.94952523,10.2357295 C7.06404332,9.7987258 6.93385572,9.33393527 6.609,9.02 L3.846,6.257 C3.65080903,6.06175007 3.65080903,5.74524993 3.846,5.55 L6.609,2.787 C6.84343098,2.5527408 6.97514738,2.23491402 6.97514738,1.9035 C6.97514738,1.57208598 6.84343098,1.2542592 6.609,1.02 Z"
                  />
                  <path
                    d="M8.954,11.624 C9.04335066,11.6434777 9.13455123,11.6532012 9.226,11.653 C9.81127096,11.6522287 10.3179139,11.2460837 10.446,10.675 L12.446,1.675 C12.5959453,0.999555759 12.1699442,0.330445338 11.4945,0.180500032 C10.8190558,0.0305547247 10.1499453,0.456555767 10,1.132 L8,10.132 C7.92859204,10.456365 7.98899504,10.7958118 8.167915,11.0756321 C8.34683495,11.3554524 8.62960831,11.5527133 8.954,11.624 L8.954,11.624 Z"
                  />
                </g>
              </g>
            </g>
          </svg>
          <svg
            aria-label="block_quote"
            class="sc-bwzfXH kVXVDT toolbar-1x1"
            height="25px"
            viewBox="0 0 18 13"
            width="25px"
          >
            <g
              fill="none"
              fill-rule="evenodd"
              stroke="none"
              stroke-width="1"
            >
              <g
                fill="#949CA2"
                fill-rule="nonzero"
                transform="translate(-581.000000, -48.000000)"
              >
                <g
                  transform="translate(581.000000, 48.000000)"
                >
                  <path
                    d="M7.44600458,0.540683605 L7.4460046,0.540683605 C3.43629213,0.653784413 0.274612689,3.93929933 0.375000251,7.88864267 L0.375000251,8.80676824 L0.375000175,8.80750625 C0.374999866,10.8235342 2.0344587,12.4578494 4.08150246,12.4578494 C6.12854622,12.4578497 7.78800474,10.8235342 7.78800474,8.80750625 C7.78800505,6.7914783 6.12854622,5.15716309 4.08150246,5.15716309 C4.08100251,5.15716309 4.08050256,5.15716319 4.0800026,5.15716339 L4.08000242,5.15716339 C3.79801349,5.15791161 3.51717657,5.19263931 3.2437519,5.26057255 L3.24375192,5.26057254 C3.14330586,5.28536873 3.0414678,5.22527619 3.01629028,5.12635189 C3.00441867,5.07970779 3.01149048,5.0303424 3.03600194,4.98875419 L3.03600178,4.98875449 C3.92604233,3.40558249 5.60871336,2.41297279 7.44600449,2.38727561 L7.44600445,2.38727561 C7.96377177,2.38727564 8.38350503,1.97390195 8.38350503,1.46397959 C8.38350505,0.954057232 7.96377177,0.540683568 7.44600445,0.540683568 L7.44600458,0.540683605 Z"
                  />
                  <path
                    d="M16.6875103,2.38727565 L16.6875102,2.38727565 C17.2052775,2.38727567 17.6250108,1.97390199 17.6250108,1.46397963 C17.6250108,0.954057269 17.2052775,0.540683605 16.6875102,0.540683605 L16.68751,0.540683611 C12.6779776,0.654185499 9.51653063,3.93946184 9.61650569,7.88864268 L9.61650569,8.80676824 L9.61650561,8.80750633 C9.61650531,10.8235343 11.2759641,12.4578495 13.3230079,12.4578495 C15.3700517,12.4578498 17.0295102,10.8235343 17.0295102,8.80750633 C17.0295105,6.79147838 15.3700517,5.15716317 13.3230079,5.15716317 C13.3227579,5.15716317 13.3225079,5.1571632 13.322258,5.15716325 L13.322258,5.15716325 C13.0400185,5.157866 12.7589281,5.19259396 12.4852574,5.2605724 L12.4852574,5.2605724 C12.4122924,5.27867912 12.3354209,5.25189054 12.2902573,5.19261781 L12.2902573,5.19261781 C12.245149,5.13323282 12.24014,5.05314176 12.2775073,4.98875405 L12.2775072,4.98875434 C13.1675477,3.40558234 14.8502187,2.41297264 16.6875099,2.38727546 L16.6875103,2.38727565 Z"
                  />
                </g>
              </g>
            </g>
          </svg>
          <svg
            aria-label="ul_list"
            class="sc-bwzfXH dOxUcK toolbar-1x2"
            height="25px"
            viewBox="0 0 16 14"
            width="25px"
          >
            <g
              fill="none"
              fill-rule="evenodd"
              stroke="none"
              stroke-width="1"
            >
              <g
                transform="translate(-582.000000, -48.000000)"
              >
                <g
                  transform="translate(580.000000, 45.000000)"
                >
                  <g>
                    <path
                      d="M3.33333333,8.75 C2.64166667,8.75 2.08333333,9.30833333 2.08333333,10 C2.08333333,10.6916667 2.64166667,11.25 3.33333333,11.25 C4.025,11.25 4.58333333,10.6916667 4.58333333,10 C4.58333333,9.30833333 4.025,8.75 3.33333333,8.75 L3.33333333,8.75 Z M3.33333333,3.75 C2.64166667,3.75 2.08333333,4.30833333 2.08333333,5 C2.08333333,5.69166667 2.64166667,6.25 3.33333333,6.25 C4.025,6.25 4.58333333,5.69166667 4.58333333,5 C4.58333333,4.30833333 4.025,3.75 3.33333333,3.75 L3.33333333,3.75 Z M3.33333333,13.8916667 C2.71666667,13.8916667 2.225,14.3916667 2.225,15 C2.225,15.6083333 2.725,16.1083333 3.33333333,16.1083333 C3.94166667,16.1083333 4.44166667,15.6083333 4.44166667,15 C4.44166667,14.3916667 3.95,13.8916667 3.33333333,13.8916667 L3.33333333,13.8916667 Z M5.83333333,15.8333333 L17.5,15.8333333 L17.5,14.1666667 L5.83333333,14.1666667 L5.83333333,15.8333333 L5.83333333,15.8333333 Z M5.83333333,10.8333333 L17.5,10.8333333 L17.5,9.16666667 L5.83333333,9.16666667 L5.83333333,10.8333333 L5.83333333,10.8333333 Z M5.83333333,4.16666667 L5.83333333,5.83333333 L17.5,5.83333333 L17.5,4.16666667 L5.83333333,4.16666667 L5.83333333,4.16666667 Z"
                      fill="#949CA2"
                    />
                    <polygon
                      points="0 0 20 0 20 20 0 20"
                    />
                  </g>
                </g>
              </g>
            </g>
          </svg>
          <svg
            aria-label="ol_list"
            class="sc-bwzfXH kVXVDT toolbar-1x3"
            height="25px"
            viewBox="0 0 17 14"
            width="25px"
          >
            <g
              fill="none"
              fill-rule="evenodd"
              stroke="none"
              stroke-width="1"
            >
              <g
                transform="translate(-617.000000, -48.000000)"
              >
                <g
                  transform="translate(616.000000, 45.000000)"
                >
                  <g>
                    <path
                      d="M1.66666667,14.1666667 L3.33333333,14.1666667 L3.33333333,14.5833333 L2.5,14.5833333 L2.5,15.4166667 L3.33333333,15.4166667 L3.33333333,15.8333333 L1.66666667,15.8333333 L1.66666667,16.6666667 L4.16666667,16.6666667 L4.16666667,13.3333333 L1.66666667,13.3333333 L1.66666667,14.1666667 L1.66666667,14.1666667 Z M2.5,6.66666667 L3.33333333,6.66666667 L3.33333333,3.33333333 L1.66666667,3.33333333 L1.66666667,4.16666667 L2.5,4.16666667 L2.5,6.66666667 L2.5,6.66666667 Z M1.66666667,9.16666667 L3.16666667,9.16666667 L1.66666667,10.9166667 L1.66666667,11.6666667 L4.16666667,11.6666667 L4.16666667,10.8333333 L2.66666667,10.8333333 L4.16666667,9.08333333 L4.16666667,8.33333333 L1.66666667,8.33333333 L1.66666667,9.16666667 L1.66666667,9.16666667 Z M5.83333333,4.16666667 L5.83333333,5.83333333 L17.5,5.83333333 L17.5,4.16666667 L5.83333333,4.16666667 L5.83333333,4.16666667 Z M5.83333333,15.8333333 L17.5,15.8333333 L17.5,14.1666667 L5.83333333,14.1666667 L5.83333333,15.8333333 L5.83333333,15.8333333 Z M5.83333333,10.8333333 L17.5,10.8333333 L17.5,9.16666667 L5.83333333,9.16666667 L5.83333333,10.8333333 L5.83333333,10.8333333 Z"
                      fill="#949CA2"
                    />
                    <polygon
                      points="0 0 20 0 20 20 0 20"
                    />
                  </g>
                </g>
              </g>
            </g>
          </svg>
          <div
            class="sc-htpNat jqGvzC toolbar-4x2"
          />
          <svg
            aria-label="link"
            class="sc-bwzfXH gBwmUW toolbar-2x1"
            height="25px"
            viewBox="0 0 14 14"
            width="25px"
          >
            <g
              fill="none"
              fill-rule="evenodd"
              stroke="none"
              stroke-width="1"
            >
              <g
                fill="#949CA2"
                fill-rule="nonzero"
                transform="translate(-727.000000, -48.000000)"
              >
                <g
                  transform="translate(727.000000, 48.000000)"
                >
                  <path
                    d="M6.10050375,10.7181733 L4.50391944,12.3141742 L4.50391947,12.3141742 C3.86064824,12.9288446 2.84768928,12.9288446 2.20441806,12.3141743 L1.6840844,11.794424 L1.68408431,11.7944239 C1.04934725,11.1595585 1.04934725,10.1303712 1.68408449,9.49550578 L5.06741991,6.11217037 L5.06741988,6.11217039 C5.70262361,5.47769 6.73171758,5.47769 7.3669213,6.11217034 L7.88783829,6.63250399 L7.88783828,6.63250399 C8.11625426,6.86091996 8.48658948,6.86091996 8.71500546,6.63250399 C8.94342143,6.40408802 8.94342143,6.03375279 8.71500546,5.80533682 C8.71500546,5.80533682 8.71500546,5.80533682 8.71500546,5.80533682 L8.19408847,5.28500316 L8.19408844,5.28500313 C7.1029686,4.19532996 5.33545584,4.19532996 4.24433601,5.2850032 L0.86100059,8.66833862 L0.861000611,8.6683386 C-0.22953006,9.7588226 -0.229576727,11.5268837 0.860907423,12.6174144 C0.860938471,12.6174454 0.860969519,12.6174765 0.861000569,12.6175075 L1.38133422,13.1372578 L1.38133403,13.1372576 C2.4707447,14.2279516 4.23806496,14.2289958 5.32875896,13.1395851 C5.32953511,13.1388099 5.3303108,13.1380342 5.33108604,13.137258 L6.92708703,11.5418404 L6.92708703,11.5418404 C7.15501942,11.314069 7.15515009,10.9446467 6.92737869,10.7167149 C6.6996073,10.4887825 6.33018499,10.4886518 6.10225319,10.7164232 C6.10225318,10.7164232 6.10225318,10.7164232 6.10225318,10.7164232 L6.10050375,10.7181733 Z"
                  />
                  <path
                    d="M13.1401748,1.37841751 L12.6198411,0.858083861 L12.6198412,0.858083983 C11.5276014,-0.228316686 9.76291196,-0.228316686 8.67067212,0.85808374 L6.97608775,2.54975145 L6.97608774,2.54975146 C6.75227485,2.78148194 6.75869385,3.15077183 6.99042404,3.3745853 C7.21647926,3.5929166 7.57486606,3.5929166 7.80091954,3.37458528 L9.49258724,1.68291757 L9.49258715,1.68291767 C10.127855,1.04914894 11.1562373,1.04914894 11.7915052,1.68291747 L12.3124222,2.20383446 L12.3124222,2.20383442 C12.9474101,2.83818065 12.9479351,3.86718128 12.3135898,4.50216917 C12.3132008,4.50255856 12.3128116,4.50294774 12.3124223,4.50333673 L8.92908684,7.88667214 L8.92908689,7.8866721 C8.62488478,8.19273795 8.21086311,8.36435472 7.77933618,8.36325572 L7.77933618,8.36325572 L7.77933619,8.36325572 C7.34826076,8.36439236 6.93465733,8.1929847 6.63075215,7.88725543 L6.63075215,7.88725543 C6.41054193,7.65209846 6.04139262,7.63998203 5.80623914,7.86019208 C5.57108217,8.08040229 5.55896574,8.4495516 5.77917579,8.68470508 C5.78761627,8.69371846 5.79634091,8.70246159 5.80533638,8.71092121 L5.8053365,8.71092133 C6.32776607,9.23599516 7.03864393,9.530091 7.77933772,9.5275885 L7.77933772,9.5275885 L7.77933771,9.5275885 C8.520224,9.5301841 9.23131361,9.23608441 9.75392226,8.71092133 L13.1372577,5.32758592 L13.1372577,5.32758593 C14.2279517,4.23772026 14.2286225,2.47002667 13.1387594,1.37933267 C13.138259,1.3788319 13.1377585,1.37833133 13.1372577,1.37783095 L13.1401748,1.37841751 Z"
                  />
                </g>
              </g>
            </g>
          </svg>
          <div
            class="sc-htpNat jqGvzC toolbar-4x3"
          />
          <svg
            aria-label="undo"
            class="sc-bwzfXH gcddOD toolbar-2x2"
            height="25px"
            viewBox="0 0 16 11"
            width="25px"
          >
            <g
              fill="none"
              fill-rule="evenodd"
              stroke="none"
              stroke-width="1"
            >
              <g
                fill="#949CA2"
                fill-rule="nonzero"
                transform="translate(-692.000000, -49.000000)"
              >
                <g
                  transform="translate(692.000000, 49.000000)"
                >
                  <path
                    d="M12.6900078,2.56651697 C10.3793397,1.21662262 7.20333777,1.40331718 3.9020024,3.06569321 L3.9020024,3.06569321 C3.83767696,3.09793767 3.75978898,3.0856688 3.70866895,3.03523948 L1.70466772,1.05309937 L1.70466772,1.05309938 C1.31280015,0.666625095 0.679640428,0.668791281 0.290460188,1.05793716 C0.104299407,1.24408355 -0.000123324148,1.49582065 -1.32407326e-06,1.75816755 L-1.32407326e-06,9.57020939 L-1.32407325e-06,9.57020949 C-1.26847808e-06,9.93584284 0.298475526,10.2322469 0.666665753,10.2322469 L8.52800392,10.2322469 L8.52800393,10.2322469 C9.08028893,10.2324955 9.52820454,9.78809065 9.52845788,9.23964095 C9.52857821,8.97599977 9.42312381,8.72313192 9.23533769,8.53676904 L7.56867,6.88167543 L7.56867001,6.88167544 C7.53082585,6.84450879 7.51322324,6.79157545 7.52133665,6.73933739 L7.52133664,6.73933741 C7.52975585,6.68706208 7.56265947,6.6418577 7.61000337,6.61752252 C9.50800453,5.65293397 11.5146724,5.49205887 12.9246733,6.24347137 C14.066674,6.84658748 14.6666744,7.99787059 14.6666744,9.56888545 L14.6666744,9.56888555 C14.6666744,9.9345189 14.9651512,10.230923 15.3333415,10.230923 C15.7015317,10.2309229 16.0000085,9.9345189 16.0000085,9.56888555 C16.0000085,6.30173076 14.8246745,3.81445608 12.6900065,2.5665155 L12.6900078,2.56651697 Z"
                  />
                </g>
              </g>
            </g>
          </svg>
          <svg
            aria-label="redo"
            class="sc-bwzfXH gcddOD toolbar-2x3"
            height="25px"
            viewBox="0 0 16 11"
            width="25px"
          >
            <g
              fill="none"
              fill-rule="evenodd"
              stroke="none"
              stroke-width="1"
            >
              <g
                fill="#949CA2"
                fill-rule="nonzero"
                transform="translate(-798.000000, -49.000000)"
              >
                <g
                  transform="translate(806.000000, 54.500000) scale(-1, 1) translate(-806.000000, -54.500000) translate(798.000000, 49.000000)"
                >
                  <path
                    d="M12.6900078,2.56651697 C10.3793397,1.21662262 7.20333777,1.40331718 3.9020024,3.06569321 L3.9020024,3.06569321 C3.83767696,3.09793767 3.75978898,3.0856688 3.70866895,3.03523948 L1.70466772,1.05309937 L1.70466772,1.05309938 C1.31280015,0.666625095 0.679640428,0.668791281 0.290460188,1.05793716 C0.104299407,1.24408355 -0.000123324148,1.49582065 -1.32407326e-06,1.75816755 L-1.32407326e-06,9.57020939 L-1.32407325e-06,9.57020949 C-1.26847808e-06,9.93584284 0.298475526,10.2322469 0.666665753,10.2322469 L8.52800392,10.2322469 L8.52800393,10.2322469 C9.08028893,10.2324955 9.52820454,9.78809065 9.52845788,9.23964095 C9.52857821,8.97599977 9.42312381,8.72313192 9.23533769,8.53676904 L7.56867,6.88167543 L7.56867001,6.88167544 C7.53082585,6.84450879 7.51322324,6.79157545 7.52133665,6.73933739 L7.52133664,6.73933741 C7.52975585,6.68706208 7.56265947,6.6418577 7.61000337,6.61752252 C9.50800453,5.65293397 11.5146724,5.49205887 12.9246733,6.24347137 C14.066674,6.84658748 14.6666744,7.99787059 14.6666744,9.56888545 L14.6666744,9.56888555 C14.6666744,9.9345189 14.9651512,10.230923 15.3333415,10.230923 C15.7015317,10.2309229 16.0000085,9.9345189 16.0000085,9.56888555 C16.0000085,6.30173076 14.8246745,3.81445608 12.6900065,2.5665155 L12.6900078,2.56651697 Z"
                  />
                </g>
              </g>
            </g>
          </svg>
        </div>
      </div>
      <div
        class="sc-ifAKCX cjbJRX"
      >
        <div>
          <div
            autocorrect="on"
            class="doc-inner"
            clausepluginprops="[object Object]"
            contenteditable="true"
            data-gramm="false"
            data-key="7"
            data-slate-editor="true"
            data-testid="editor"
            editorprops="[object Object]"
            role="textbox"
            spellcheck="true"
            style="outline: none; white-space: pre-wrap; word-wrap: break-word;"
          >
            <p
              data-key="8"
              data-slate-object="block"
            >
              <span
                data-key="9"
                data-slate-object="text"
              >
                <span
                  data-offset-key="9:0"
                  data-slate-leaf="true"
                >
                  <span
                    data-slate-string="true"
                  >
                    Welcome! Edit this text to get started.
                  </span>
                </span>
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ui popup transition visible"
    style="left: 0px; position: absolute; top: 0px; opacity: 0; pointer-events: none; z-index: -1;"
  >
    <form
      class="ui form"
    >
      <div
        class="field"
      >
        <label>
          Link Text
        </label>
        <div
          class="ui input"
        >
          <input
            name="text"
            placeholder="Text"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <label>
          Link URL
        </label>
        <div
          class="ui input"
        >
          <input
            name="url"
            placeholder="http://example.com"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <button
          class="ui secondary disabled right floated button"
          disabled=""
          tabindex="-1"
        >
          Remove
        </button>
        <button
          class="ui primary right floated button"
          type="submit"
        >
          Apply
        </button>
      </div>
    </form>
  </div>
</body>
`;
